libbeat/keystore/file_keystore.go

```diff
@@ -0,0 +1,383 @@
+func (k *FileKeystore) Retrieve(key string) (*SecureString, error) {
+	k.RLock()
+	defer k.RUnlock()
+
+	secret, ok := k.secrets[key]
+	if !ok {
+		return nil, ErrKeyDoesntExists
+	}
+	return NewSecureString(secret.Value), nil
+}
+
+// Store add the key pair to the secret store and mark the store as dirty.
+func (k *FileKeystore) Store(key string, value []byte) error {
+	k.Lock()
+	defer k.Unlock()
+
+	k.secrets[key] = serializableSecureString{Value: value}
+	k.dirty = true
+	return nil
+}
+
+// Delete an existing key from the store and mark the store as dirty.
+func (k *FileKeystore) Delete(key string) error {
+	k.Lock()
+	defer k.Unlock()
+
+	delete(k.secrets, key)
+	k.dirty = true
+	return nil
+}
+
+// Save persists the in memory data to disk if needed.
+func (k *FileKeystore) Save() error {
+	k.Lock()
+	defer k.Unlock()
+
+	if k.dirty == false {
+		return nil
+	}
+
+	temporaryPath := fmt.Sprintf("%s.tmp", k.Path)
+	if _, err := os.Stat(temporaryPath); os.IsExist(err) {
+		return fmt.Errorf("temporary keystore exist at '%s', delete it and try again", temporaryPath)
+	}
+
+	w := new(bytes.Buffer)
+	jsonEncoder := json.NewEncoder(w)
+	if err := jsonEncoder.Encode(k.secrets); err != nil {
+		return fmt.Errorf("cannot serialize the keystore before saving it to disk: %v", err)
+	}
+
+	encrypted, err := k.encrypt(w)
+	if err != nil {
+		return fmt.Errorf("cannot encrypt the keystore: %v", err)
+	}
+
+	f, err := os.OpenFile(temporaryPath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, filePermission)
+
+	if err != nil {
+		return fmt.Errorf("cannot open file to save the keystore to '%s', error: %s", k.Path, err)
+	}
+
+	defer f.Close()
+
+	f.Write(version)
+
+	base64Encoder := base64.NewEncoder(base64.StdEncoding, f)
+	io.Copy(base64Encoder, encrypted)
+	base64Encoder.Close()
+
+	err = os.Rename(temporaryPath, k.Path)
+
+	if err != nil {
+		os.Remove(temporaryPath)
+		return fmt.Errorf("cannot replace the existing keystore, with the new keystore file at '%s', error: %s", k.Path, err)
+	}
+
+	os.Remove(temporaryPath)
+
+	k.dirty = false
+	return nil
+}
+
+// List return the availables keys.
+func (k *FileKeystore) List() ([]string, error) {
+	k.RLock()
+	defer k.RUnlock()
+
+	keys := make([]string, 0, len(k.secrets))
+
+	for key := range k.secrets {
+		keys = append(keys, key)
+	}
+
+	return keys, nil
+}
+
+// GetConfig returns common.Config representation of the key / secret pair to be merged with other
+// loaded configuration.
+func (k *FileKeystore) GetConfig() (*common.Config, error) {
+	k.RLock()
+	defer k.RUnlock()
+
+	configHash := make(map[string]interface{})
+
+	for key, secret := range k.secrets {
+		configHash[key] = string(secret.Value)
+	}
+
+	return common.NewConfigFrom(configHash)
+}
+
+// Create create an empty keystore, if the store already exist we will return an error.
+func (k *FileKeystore) Create(override bool) error {
+	k.Lock()
+
+	_, err := os.Stat(k.Path)
+
+	if os.IsExist(err) && !override {
```

Comment:
Same issue here with the logic for testing if a file exists. I suggest to not check it here at all. Just call `Save()` and use the value of the `override` to determine what flags you pass to `os.OpenFile`. Like 

```
flags := os.O_RDWR | os.O_CREAT
if override {
    flags |= os.O_TRUNC
} else { 
    flags |= os.O_EXCL
}
````